src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,343 @@ public async Task TestRemoteWorkspaceCircularReferences()
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(Guid.NewGuid().ToString()))),
+                ImmutableArray<(string, SourceText)>.Empty);
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree12(bool syncWithRemoteServer)
+        {
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray<(string, SourceText)>.Empty,
+                ImmutableArray.Create(("SG.cs", CreateText(Guid.NewGuid().ToString()))));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree13(bool syncWithRemoteServer)
+        {
+            var contents = Guid.NewGuid().ToString();
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(contents))),
+                ImmutableArray.Create(("SG.cs", CreateText(contents)), ("SG1.cs", CreateText(contents))));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree14(bool syncWithRemoteServer)
+        {
+            var contents = Guid.NewGuid().ToString();
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(contents))),
+                ImmutableArray.Create(("SG.cs", CreateText(contents)), ("SG1.cs", CreateText("Other"))));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree15(bool syncWithRemoteServer)
+        {
+            var contents = Guid.NewGuid().ToString();
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(contents))),
+                ImmutableArray.Create(("SG1.cs", CreateText(contents)), ("SG.cs", CreateText("Other"))));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree16(bool syncWithRemoteServer)
+        {
+            var contents = Guid.NewGuid().ToString();
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(contents))),
+                ImmutableArray.Create(("SG1.cs", CreateText("Other")), ("SG.cs", CreateText(contents))));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task InProcAndRemoteWorkspaceAgree17(bool syncWithRemoteServer)
+        {
+            var contents = Guid.NewGuid().ToString();
+            await TestInProcAndRemoteWorkspace(
+                syncWithRemoteServer,
+                ImmutableArray.Create(("SG.cs", CreateText(contents))),
+                ImmutableArray.Create(("SG1.cs", CreateText("Other")), ("SG.cs", CreateText(contents))),
+                ImmutableArray<(string, SourceText)>.Empty);
```

Comment:
easy pattern for just simulating a series of SG generation steps.  easy to add more tests in the future.